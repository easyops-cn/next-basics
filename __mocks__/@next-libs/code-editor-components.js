import React from "react";

export function CodeEditorItem() {
  return <div>code editor</div>;
}
