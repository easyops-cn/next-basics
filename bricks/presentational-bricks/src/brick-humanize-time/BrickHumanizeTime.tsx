import React from "react";
import {
  humanizeTime,
  HumanizeTimeFormat,
  costTime,
} from "@next-libs/datetime";
import { BrickLink } from "../brick-link/BrickLink";
import moment from "moment";
import { isNil } from "lodash";

interface BrickHumanizeTimeProps {
  value: number | string;
  isMicrosecond?: boolean;
  formatter?: HumanizeTimeFormat;
  isCostTime?: boolean;
  inputFormat?: string;
  outputFormat?: string;
  link?: {
    url: string;
    target?: string;
  };
}

export function BrickHumanizeTime(
  props: BrickHumanizeTimeProps
): React.ReactElement {
  const {
    value,
    isMicrosecond,
    inputFormat,
    outputFormat,
    isCostTime,
    formatter,
    link,
  } = props;

  if (isNil(value)) {
    return <span>-</span>;
  }

  let ts;
  if (typeof value === "number") {
    ts = isMicrosecond ? value : Number(value) * 1000;
  } else {
    const time = moment(value, inputFormat);
    ts = time.unix() * 1000;
  }

  let label: string;
  if (outputFormat) {
    label = moment(ts).format(outputFormat);
  } else {
    label = isCostTime
      ? costTime(ts)
      : humanizeTime(ts, HumanizeTimeFormat[formatter]);
  }

  if (link) {
    return <BrickLink url={link.url} target={link.target} label={label} />;
  }

  return <span>{label}</span>;
}
