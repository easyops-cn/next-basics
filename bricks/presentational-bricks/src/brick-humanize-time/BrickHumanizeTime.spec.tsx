import React from "react";
import { shallow } from "enzyme";
import { BrickHumanizeTime } from "./BrickHumanizeTime";
import { HumanizeTimeFormat } from "@next-libs/datetime";
import { install, InstalledClock } from "lolex";

describe("BrickHumanizeTime", () => {
  let clock: InstalledClock;
  beforeEach(() => {
    clock = install({ now: +new Date("2019-10-17 17:20:00") });
  });
  afterEach(() => {
    clock.uninstall();
  });

  it("should work", () => {
    const wrapper = shallow(
      <BrickHumanizeTime
        value={1563506779}
        formatter={HumanizeTimeFormat.accurate}
      />
    );
    expect(wrapper).toMatchSnapshot();
    wrapper.setProps({ value: 0 });
    wrapper.update();
    expect(wrapper.find("span").text()).toBe("January 1, 1970 08:00");
    wrapper.setProps({ value: null });
    wrapper.update();
    expect(wrapper.find("span").text()).toBe("-");
  });

  it("should work, isCostTime", () => {
    const wrapper = shallow(
      <BrickHumanizeTime
        value={1563509000}
        isMillisecond={true}
        isCostTime={true}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("should work, link props", () => {
    const wrapper = shallow(
      <BrickHumanizeTime value={1563509} link={{ url: "/a/b/c" }} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("inputFormat and outputFormat should work", () => {
    const wrapper = shallow(
      <BrickHumanizeTime value="2020-02-27 17:18" outputFormat="YYYY-MM-DD" />
    );
    expect(wrapper.find("span").first().text()).toBe("2020-02-27");
  });
});
