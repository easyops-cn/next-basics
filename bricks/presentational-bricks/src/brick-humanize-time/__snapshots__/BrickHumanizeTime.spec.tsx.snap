// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrickHumanizeTime should work 1`] = `
<span>
  7/19/2019ib19-5amt5ti265:FOR7AMT_011ORT_19AM2019
</span>
`;

exports[`BrickHumanizeTime should work, isCostTime 1`] = `
<span>
  18 libs-datetime:DARS 2 libs-datetime:HOURS
</span>
`;

exports[`BrickHumanizeTime should work, link props 1`] = `
<BrickLink
  label="January 19, 1970 am10:18"
  url="/a/b/c"
/>
`;
